Mathlib/Geometry/Manifold/Algebra/SmoothGerm.lean

```diff
@@ -0,0 +1,127 @@
+  rfl
+
+end definition
+
+-- If `R` has the appropriate structure, `smoothGerm I I' R x` is a subgroup etc.
+-- All axioms are easy to prove by choosing explicit representatives.
+section subring
+
+-- xxx: would this be useful? HasZero resp. HasOne imply the space of smooth germs has one...
+variable (I' : ModelWithCorners 𝕜 E' H')
+  (R : Type*) [CommRing R] [TopologicalSpace R] [ChartedSpace H' R]
+
+/-- If `R` is a manifold with smooth multiplication,
+`smoothGerm I I' R x` is a sub-semigroup of `Germ (𝓝 x) R`. -/
+def smoothGerm.toSubsemigroup [SmoothMul I' R] (x : M) : Subsemigroup (Germ (𝓝 x) R) where
+  carrier := smoothGerm I I' R x
+  mul_mem' ha hb := by
+    choose f hf using ha
+    choose g hg using hb
+    exact ⟨f * g, by rw [← hf, ← hg, SmoothMap.coe_mul, Germ.coe_mul]⟩
+
+/-- If `R` is a manifold with smooth multiplication,
+`smoothGerm I I' R x` is a submonoid of `Germ (𝓝 x) R`. -/
+-- FIXME: is this definition useful, given it has the same assumptions has `toSubsemigroup`?
+def smoothGerm.toSubmonoid [SmoothMul I' R] (x : M) : Submonoid (Germ (𝓝 x) R) where
+  toSubsemigroup := smoothGerm.toSubsemigroup I I' R x
+  one_mem' := ⟨1, by rw [SmoothMap.coe_one, Germ.coe_one]⟩
+
+/-- If `R` is a manifold with smooth addition,
+`smoothGerm I I' R x` is an additive sub-semigroup of `Germ (𝓝 x) R`. -/
+def smoothGerm.toAddSubsemigroup [SmoothAdd I' R] (x : M) : AddSubsemigroup (Germ (𝓝 x) R) where
+  carrier := smoothGerm I I' R x
+  add_mem' ha hb := by
+    choose f hf using ha
+    choose g hg using hb
+    exact ⟨f + g, by rw [← hf, ← hg, SmoothMap.coe_add, Germ.coe_add]⟩
+
+/-- If `G` is an additive Lie group, `smoothGerm I I' G x` is
+  an additive subgroup of `Germ (𝓝 x) G`. -/
+def smoothGerm.toAddSubgroup [LieAddGroup I' R] (x : M) : AddSubgroup (Germ (𝓝 x) R) where
+  __ := smoothGerm.toAddSubsemigroup I I' R x
+  zero_mem' := ⟨0, by rw [SmoothMap.coe_zero, Germ.coe_zero]⟩
+  neg_mem' h := by
+    choose f hf using h
+    exact ⟨-f, by rw [← hf, SmoothMap.coe_neg, Germ.coe_neg]⟩
+
+/-- If `R` is a smooth ring, `smoothGerm I I' R x` is a subring of `Germ (𝓝 x) R`. -/
+def smoothGerm.toSubring [SmoothRing I' R] (x : M) : Subring (Germ (𝓝 x) R) where
+  __ := smoothGerm.toSubmonoid I I' R x
+  __ := smoothGerm.toAddSubgroup I I' R x
+
+-- xxx: is this lemma useful?
+lemma smoothGerm.toSubring_mem_coe [SmoothRing I' R] {x : M} (a : Germ (𝓝 x) R) :
+    a ∈ smoothGerm.toSubring I I' R x ↔ a ∈ smoothGerm I I' R x := Iff.rfl
+
+/-- The map `C^∞(M, R) → Germ (𝓝 x) R` as a ring homomorphism, for a smooth ring `R`. -/
+def RingHom.germOfContMDiffMap (R : Type*) [CommRing R] [TopologicalSpace R] [ChartedSpace H' R]
+    [SmoothRing I' R] (x : M) : C^∞⟮I, M; I', R⟯ →+* Germ (𝓝 x) R :=
+  (Germ.coeRingHom _).comp SmoothMap.coeFnRingHom
+
+lemma toSubring_eq_range [SmoothRing I' R] (x : M) :
+    smoothGerm.toSubring I I' R x = (RingHom.germOfContMDiffMap I I' R x).range := by
+  rfl
```

Comment:
Namespaces are all over the place here. Can you maybe put everything in namespace `smoothGerm`? I also don't think that there needs to be `RingHom` in the name of `RingHom.germOfContMDiffMap`.

Reply:
Good point, addressed!